datafusion/physical-expr/src/expressions/cast.rs

```diff
@@ -93,6 +94,36 @@ impl PhysicalExpr for CastExpr {
         let value = self.expr.evaluate(batch)?;
         cast_column(&value, &self.cast_type, &self.cast_options)
     }
+
+    fn children(&self) -> Vec<Arc<dyn PhysicalExpr>> {
+        vec![self.expr.clone()]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        children: Vec<Arc<dyn PhysicalExpr>>,
+    ) -> Result<Arc<dyn PhysicalExpr>> {
+        Ok(Arc::new(CastExpr::new(
+            children[0].clone(),
+            self.cast_type.clone(),
+            CastOptions {
+                safe: self.cast_options.safe,
+            },
+        )))
+    }
+}
+
+impl PartialEq<dyn Any> for CastExpr {
+    fn eq(&self, other: &dyn Any) -> bool {
+        down_cast_any_ref(other)
+            .downcast_ref::<Self>()
+            .map(|x| {
+                self.expr.eq(&x.expr)
+                    && self.cast_type == x.cast_type
+                    && self.cast_options.safe == x.cast_options.safe
```

Comment:
This seems to be a bit error-prone, considering that `cast_options` might get new members and this part will not count them. Can we consider adding equality comparison support to `CastOptions` directly?

Reply:
Nevermind, it seems like we can't change it here. Created a ticket on arrow-rs.